Mathlib/Analysis/Fourier/ZMod.lean

```diff
@@ -0,0 +1,67 @@
+/-
+Copyright (c) 2024 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.Analysis.SpecialFunctions.Complex.Circle
+import Mathlib.Analysis.Fourier.FourierTransform
+import Mathlib.NumberTheory.DirichletCharacter.GaussSum
+
+/-!
+# Fourier theory on `ZMod N`
+
+Basic definitions and properties of the discrete Fourier transform for functions on `ZMod N`.
+
+### Main definitions and results
+
+* `ZMod.dft`: the Fourier transform with respect to the standard additive character
+  `ZMod.stdAddChar` (mapping `j mod N` to `exp (2 * π * I * j / N)`). The notation `𝓕`, scoped in
+  namespace `ZMod`, is available for this.
+* `DirichletCharacter.fourierTransform_eq_inv_mul_gaussSum`: the discrete Fourier transform of a
+  primitive Dirichlet character `χ` is a Gauss sum times `χ⁻¹`.
+-/
+
+open scoped Real
+
+open MeasureTheory
+
+namespace ZMod
+
+variable {N : ℕ} [NeZero N]
+
+/-- The discrete Fourier transform on `ℤ / N ℤ` (with the counting measure) -/
+noncomputable def dft (Φ : ZMod N → ℂ) (k : ZMod N) : ℂ :=
+  Fourier.fourierIntegral toCircle Measure.count Φ k
+
+@[inherit_doc] scoped notation "𝓕" => dft
+
+lemma dft_apply (Φ : ZMod N → ℂ) (k : ZMod N) :
+    𝓕 Φ k = ∑ j : ZMod N, toCircle (-(j * k)) • Φ j := by
+  simp only [dft, Fourier.fourierIntegral_def, integral_countable' <| .of_finite ..,
+    Measure.count_singleton, ENNReal.one_toReal, one_smul, tsum_fintype]
+
+lemma dft_def (Φ : ZMod N → ℂ) : 𝓕 Φ = fun k ↦ ∑ j : ZMod N, toCircle (-(j * k)) • Φ j :=
+  funext (dft_apply Φ)
+
+end ZMod
+
+open ZMod
+
+namespace DirichletCharacter
+
+variable {N : ℕ} [NeZero N] (χ : DirichletCharacter ℂ N)
+
+lemma fourierTransform_eq_gaussSum_mulShift (k : ZMod N) :
+    𝓕 χ k = gaussSum χ (stdAddChar.mulShift (-k)) := by
+  simp only [dft_def]
+  refine congr_arg Finset.univ.sum (funext fun j ↦ ?_)
```

Comment:
```suggestion
  refine Fintype.sum_congr rfl fun j ↦ ?_
```

Reply:
I'm afraid your suggestion does not work. (It took me a while to spot since there's a lot of library to compile in between `Topology.Instances.AddCircle` and here.) 

I guess you meant either `Finset.sum_congr rfl fun j _ ↦ ?_` or `Fintype.sum_congr _ _ fun j ↦ ?_`. However, while playing with this I spotted that `congr 1 with j` also works, and IMHO that's nicer than any of these other possibilities, so I am going with that.